### PR TITLE
feat: Add ability to upload a plugin via URL [JENKINS-4814]

### PR DESCRIPTION
**Pull request by slide**
See [JENKINS-4814](https://issues.jenkins-ci.org/browse/JENKINS-4814).

- Adds input element for a URL to an hpi/jpi file
- Processes the URL or File upload via PluginCopier interface implementations
- Adds "accept" attribute to file input item to restrict file types for upload

No tests are added, there are currently none for the doUploadPlugin as far as I could see.

I'm sure I am missing something here, but this is something that I have wanted. In the case of a bad url there is currently an exception thrown and the exception page shown, which is not ideal, but I am not sure what would be better in this case.

### Proposed changelog entries

* Accept plugin upload by URL in addition to file name.

### Proposed upgrade guidelines

* N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). 
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

Unsure

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — COMMENTED** [review]
nice PR!

I've tested it out and it works, one concern on UI.

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:74]: I made some small modifications to the UI, I don't think they address you concerns about the buttons but they simplify a little bit

image

File: core/src/main/java/hudson/PluginManager.java
Comment (slide) [reply core/src/main/java/hudson/PluginManager.java:1770]: I added a check method that will show a warning if you are using a non-https url, is that ok, or do you want something more explicit for non-http/https?

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: > the UX here isn't the best imo.
> 
> I think either:
> 
> * an expandable radio section where you select what type and a field pops out
> * just a radio option and a single field, but depends on messaging.
> 
> radio options would be:
> 
> * File
> * URL
> 
> the main issue I think is soo many primary buttons so close together:
> 
> image
> 
> and also these are the same thing even is the source is different so I don't think it justifies another form.
> 
> any thoughts @fqueiruga @janfaracik @uhafner?

Yes, from a user experience view it would make more sense to have one "Deploy" button with two fields. One for the file and one for the URL. I don't think that an additional radio button is required. 

(But this view and the upload concept is also in a bad UI shape yet so it will not make it dramatically difference)

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: I had it originally as just two fields but @daniel-beck asked me to separate into two sections. Perhaps it could be combined into one section and the text "Upload Plugin" could be changed to "Deploy Plugin" with the two fields?

Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Yes, that would be better. I wonder if it would be even better if we would add an additional tab for the upload. So we would have 
- Updates
- Available
- Installed
- Deploy
- Advanced (or better Settings?)

Then we can use the whole form with a single primary button for the upload.

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Here is what a combined section would look like:

image

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Here is what a combined section would look like:

image

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: The text under deploy plugin needs changing from upload, I would change it to say plugin instead of `.hpi`.

between the file and url fields what about adding a `<p>Or</p>`

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Something like this? 
image

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Much nicer, not sure if those - are needed

Comment (Wadeck) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: (Personal opinion) I really prefer the approach with the "or" instead of having to click on a radio to select the type of input. This adds a non-necessary click I would say :)

Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:68]: Yes, this looks good!

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: @timja I wasn't sure about the - either, but have seen it in other places on the web. Here is the current commit: 
image

Comment (slide) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly]: @timja I wasn't sure about the - either, I had seen it in other places on the web, but I think it looks better without it. Here is what it looks like with the current commit:

image

File: core/src/main/java/hudson/PluginManager.java
Comment (timja) [inline core/src/main/java/hudson/PluginManager.java:1795]: is it safe to do a `items.get(1)` without any length checks?

File: core/src/main/resources/hudson/PluginManager/advanced.properties
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/advanced.properties:23]: remove translations that are no longer used now please

File: core/src/main/java/hudson/PluginManager.java
Comment (slide) [reply core/src/main/java/hudson/PluginManager.java:1795]: Previously the code was doing .get(0) with no length check. Each form field creates and item and then there are any crumbs, etc. I was thinking about reworking to use stream and look for the field names, but followed the existing method.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
@slide you up for updating the docs? https://www.jenkins.io/doc/book/managing/plugins/#from-the-web-ui-2

cc @MarkEWaite

**Comment by slide** [comment]
> @slide you up for updating the docs? https://www.jenkins.io/doc/book/managing/plugins/#from-the-web-ui-2
> 
> cc @MarkEWaite

Sure, I'll take a look

**Comment by slide** [comment]
See https://github.com/jenkins-infra/jenkins.io/pull/4677

File: core/src/main/java/hudson/PluginManager.java
Comment (timja) [reply core/src/main/java/hudson/PluginManager.java:1771]: did you find this out @slide ?

Comment (slide) [reply core/src/main/java/hudson/PluginManager.java:1771]: No, I did not. Underneath, the class is just using a call to URL.openConnection

**Review by uhafner — APPROVED** [review]
UI is looking good now! 🎉 

(We still missed the opportunity to split this *advanced* section in two disjunct sections *deploy* and *proxy* with a single *primary* button on each, but this will be a good Hacktoberfest issue for the next year 😉)

**Comment by timja** [comment]
All 3 sections should be in different places, and proxy has no business on that page, but we can deal with it at some point.

I’m excited for this feature, will be great for things like core-pr-tester with incremental builds, saves me downloading it

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
